### PR TITLE
Support resolve reference for typeof(object) on deserialize

### LINKED ISSUE #1776: Preserved references should avoid calling object converter
As of now, if the JSON object that is passed to the internal converter for an `object` type is a reference `{ "$ref": "#" }` we create a `JsonElement` for such object/property.

Instead we could search into the map of references and take the value from there.

Repro:
```cs
private class Node
{
    public Node Child { get; set; }
    public object Sibling { get; set; }
}

[Fact]
public static void PreserveReferenceOnTypesWithObjectConverter()
{
    Node root = new Node();
    root.Child = new Node();
    root.Sibling = root.Child;
    // Ok.
    Assert.Same(root.Child, root.Sibling);

    JsonSerializerOptions opts = new JsonSerializerOptions
    {
        ReferenceHandling = ReferenceHandling.Preserve
    };

    string json = JsonSerializer.Serialize(root, opts);
    /* serialized object:
    {
        "$id": "1",
        "Child": {
            "$id": "2"
        },
        "Sibling": {
            "$ref": "2"
        }
    }
    */

    Node rootCopy = JsonSerializer.Deserialize<Node>(json, opts);
    // Fails, rootCopy.Sibling is a JsonElement.
    Assert.Same(rootCopy.Child, rootCopy.Sibling);
}
```

cc @ahsonkhan, @steveharter, @layomia

### PR DESCRIPTION
**Pull request by jozkee**
Fixes #1776

Analyze the `JsonElement` to be returned, determines if it can be interpreted as a reference (`{"$ref":"1"}`), lookup for the reference and returns it instead.

This is performed to enable round-tripping on scenarios where a property that is `typeof(object)` is a reference to an instance of another type e.g:
```cs
private class Node
{
    public Node Child { get; set; }
    public object Sibling { get; set; }
}
```

This change also grants parity with Json.NET in said scenario.

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs [hunk 1]
```diff
@@ -271,5 +274,46 @@ internal static MetadataPropertyName GetMetadataPropertyName(ReadOnlySpan<byte>
...
+                int propertyCount = 0;
+                foreach (JsonProperty property in element.EnumerateObject())
+                {
+                    propertyCount++;
+                    if (refMetadataFound)
+                    {
+                        // There are more properties in an object with $ref.
+                        ThrowHelper.ThrowJsonException_MetadataReferenceObjectCannotContainOtherProperties();
+                    }
+                    else if (property.EscapedNameEquals(s_refPropertyName))
```
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs:297]: Looks like we lose the validation that the metadata in the payload was the unescaped literal "$ref". Do we care?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs [hunk 2]
```diff
@@ -150,6 +150,19 @@ internal bool TryRead(ref Utf8JsonReader reader, Type typeToConvert, JsonSeriali
                         ref reader);
                 }
 
+                if (CanBePolymorphic && options.ReferenceHandler != null)
```
Comment (layomia) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:153]: Could another approach be to intercept the call to the Object/JsonElementConverter (when the ref handling feature is on) and just read the $ref metadata directly?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs (hunk 1 above)
Comment (Jozkee) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs:297]: Hmmm not really, unlike `NameEquals`; `EscapedNameEquals` will compare against the raw bytes, i.e. it won't unescape the name.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs (hunk 2 above)
Comment (Jozkee) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:153]: Yes, but we would have to create a copy of the reader that we would need to throw away if the payload doesn't met the required conditions (being an object, contain only the $ref property). Since this is an edge case, I think this approach is cleaner.

File: src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs [hunk 3]
```diff
@@ -60,21 +60,21 @@ private static async Task PerformSerialization<TElement>(object obj, Type type,
...
-                await JsonSerializer.SerializeAsync(memoryStream, obj, options);
-                string serialized = Encoding.UTF8.GetString(memoryStream.ToArray());
-                JsonTestHelper.AssertJsonEqual(expectedjson, serialized);
+            using var memoryStream = new MemoryStream();
+            await JsonSerializer.SerializeAsync(memoryStream, obj, options);
+            string serialized = Encoding.UTF8.GetString(memoryStream.ToArray());
+            JsonTestHelper.AssertJsonEqual(expectedjson, serialized);
 
-                memoryStream.Position = 0;
-                await TestDeserialization<TElement>(memoryStream, expectedjson, type, options);
```
Comment (layomia) [inline src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs:70]: Why was this line deleted?

File: src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs [hunk 4]
```diff
@@ -353,6 +353,16 @@ private static IEnumerable<Type> DictionaryTypes<TElement>()
             typeof(GenericIReadOnlyDictionaryWrapper<string, TElement>)
         };
 
+        // Non-generic types cannot roundtrip when they contain a $ref written on serialization and they are the root type.
```
Comment (layomia) [inline src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs:356]: What's a minimal code sample to highlight this behavior?

**Review by layomia — APPROVED** [review]

File: src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs (hunk 3 above)
Comment (Jozkee) [reply src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs:70]: It is still there on line 72, I added the condition `if (options.ReferenceHandler == null || !GetTypesNonRoundtrippableWithReferenceHandler().Contains(type))` to skip it given that it won't work if the root type uses `object` type elements and has a $ref object.

File: src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs (hunk 4 above)
Comment (Jozkee) [reply src/libraries/System.Text.Json/tests/Serialization/Stream.Collections.cs:356]: key1 will be parsed as a `JsonElement` and when key2 tries to look for the reference, it will not find it.
```cs
using System;
using System.Collections;
using System.Text.Json;
using System.Text.Json.Serialization;

public class Program
{
  public static void Main()
  {
    var foo = new Foo();
    Hashtable h = new Hashtable();
    h.Add("key1", foo);
    h.Add("key2", foo);

    var opts = new JsonSerializerOptions();
    opts.ReferenceHandler = ReferenceHandler.Preserve;

    string json = JsonSerializer.Serialize(h, opts);
    Console.WriteLine(json);
    // {"$id":"1","key1":{"$id":"2"},"key2":{"$ref":"2"}}

    Hashtable copyOfh = JsonSerializer.Deserialize<Hashtable>(json, opts);
  }
}
```
